wgpu/src/buffers/buffer.rs

```diff
@@ -0,0 +1,91 @@
+//! Utilities for static buffer operations.
+
+/// A generic buffer struct useful for items which have no alignment requirements
+/// (e.g. Vertex, Index buffers) and are set once and never changed until destroyed.
+///
+/// This buffer is mapped to the GPU on creation, so must be initialized with the correct capacity.
+#[derive(Debug)]
+pub(crate) struct StaticBuffer {
+    //stored sequentially per mesh iteration
+    offsets: Vec<wgpu::BufferAddress>,
+    gpu: wgpu::Buffer,
+    //the static size of the buffer
+    size: wgpu::BufferAddress,
+}
+
+impl StaticBuffer {
+    pub fn new(
+        device: &wgpu::Device,
+        label: &'static str,
+        size: u64,
+        usage: wgpu::BufferUsages,
+        total_offsets: usize,
+    ) -> Self {
+        Self {
+            offsets: Vec::with_capacity(total_offsets),
+            gpu: device.create_buffer(&wgpu::BufferDescriptor {
+                label: Some(label),
+                size,
+                usage,
+                mapped_at_creation: true,
+            }),
+            size,
+        }
+    }
+
+    /// Resolves pending write operations & unmaps buffer from host memory.
+    pub fn flush(&self) {
+        (&self.gpu).unmap();
+    }
+
+    /// Returns whether or not the buffer needs to be recreated. This can happen whenever the mesh 
+    /// data is re-submitted.
+    pub fn needs_recreate(&self, new_size: usize) -> bool {
+        self.size != new_size as u64
+    }
+
+    /// Writes the current vertex data to the gpu buffer with a memcpy & stores its offset.
+    pub fn write(&mut self, offset: u64, content: &[u8]) {
```

Comment:
If we have a generic type `T` we can ask for a slice of `&[T]` here and require the `bytemuck` traits to convert the slice to bytes in this method.